src/on-chain/OnChainSurplus.py

```diff
@@ -0,0 +1,135 @@
+
+class DecodedSettlement:
+    def __init__(self, tokens: List[str], clearing_prices: List[int], trades: List[Tuple[int, int, str, int, int, int, bytes, int, int, int, bytes]], interactions: List[List[Tuple[str, int, bytes]]]):
+        self.tokens = tokens
+        self.clearing_prices = clearing_prices
+        self.trades = trades
+        self.interactions = interactions
+
+    @classmethod
+    def new(cls, contractInstance, transaction):
+        # Decode the function input
+        decoded_input = contractInstance.decode_function_input(transaction)[1:]
+        # Convert the decoded input to the expected types
+        tokens = decoded_input[0]['tokens']
+        clearing_prices = decoded_input[0]['clearingPrices']
+        trades = decoded_input[0]['trades']
+        interactions = decoded_input[0]['interactions']
+
+        # Create and return a new instance of DecodedSettlement
+        return cls(tokens, clearing_prices, trades, interactions)
+
+
+def fetchSettlements(startBlock = None, endBlock = None, settlementHash = None, w3 = None, contractInstance = None):
+    transactionHashes = []
+    if settlementHash is not None:
+        transactionHashes.append(settlementHash)
+
+    elif startBlock is not None and endBlock is not None:
+    # define the filter criteria
+        filter_criteria = {
+            'fromBlock': int(startBlock),
+            'toBlock': int(endBlock),
+            'address': address,
+        }
+        transactions = w3.eth.filter(filter_criteria).get_all_entries()
+        for tx in transactions:
+            txHash = (tx['transactionHash']).hex()
+            if txHash not in transactionHashes:
+                transactionHashes.append(txHash)
+        transactionHashes.reverse()
+
+        # At this point we have all the needed hashes
+    decodeHashes(transactionHashes, w3, contractInstance)
+
+
+def decodeHashes(transactionHashes, w3, contractInstance):
+    for hashCounter in range(len(transactionHashes)):
+        singleHash = transactionHashes[hashCounter]
+        decoder(singleHash, w3, contractInstance)
+
+
+def decoder(settlementHash, w3, contractInstance):
+    encodedTransaction = w3.eth.get_transaction(settlementHash)
+    encodedTransaction = encodedTransaction.input
+    decoded_settlement = DecodedSettlement.new(contractInstance, encodedTransaction)
+
+    endpointDataJson = requests.get(f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{settlementHash}")
+    endpointDataPy = json.loads(endpointDataJson.text)
+    auctionId = endpointDataPy["auctionId"]
+    bucketResponseJson = requests.get(f"https://solver-instances.s3.eu-central-1.amazonaws.com/prod-mainnet/{auctionId}.json")
+    bucketResponsePy = json.loads(bucketResponseJson.text)
+
+    # there can be multiple orders/trades in a single settlement
+    for ordersCount in range(len(decoded_settlement.trades)):
+        trade = decoded_settlement.trades[ordersCount]
+        sellTokenIndex = trade['sellTokenIndex']
```

Comment:
Ran it locally and this step doesn't seem to work. Maybe there is an issue with the decoding?